ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.manager.peertask;
+
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.p2p.rlpx.connections.PeerConnection.PeerNotConnected;
+import org.hyperledger.besu.ethereum.p2p.rlpx.wire.MessageData;
+import org.hyperledger.besu.metrics.BesuMetricCategory;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import org.hyperledger.besu.plugin.services.metrics.Counter;
+import org.hyperledger.besu.plugin.services.metrics.LabelledGauge;
+import org.hyperledger.besu.plugin.services.metrics.LabelledMetric;
+import org.hyperledger.besu.plugin.services.metrics.OperationTimer;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/** Manages the execution of PeerTasks, respecting their PeerTaskRetryBehavior */
+public class PeerTaskExecutor {
+
+  private final PeerSelector peerSelector;
+  private final PeerTaskRequestSender requestSender;
+
+  private final LabelledMetric<OperationTimer> requestTimer;
+  private final LabelledMetric<Counter> partialSuccessCounter;
+  private final LabelledMetric<Counter> timeoutCounter;
+  private final LabelledMetric<Counter> invalidResponseCounter;
+  private final LabelledMetric<Counter> internalExceptionCounter;
+  private final LabelledGauge inflightRequestGauge;
+  private final Map<String, AtomicInteger> inflightRequestCountByClassName;
+
+  public PeerTaskExecutor(
+      final PeerSelector peerSelector,
+      final PeerTaskRequestSender requestSender,
+      final MetricsSystem metricsSystem) {
+    this.peerSelector = peerSelector;
+    this.requestSender = requestSender;
+    requestTimer =
+        metricsSystem.createLabelledTimer(
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:RequestTime",
+            "Time taken to send a request and receive a response",
+            "className");
+    partialSuccessCounter =
```

Comment:
For counters the convention we are using generally and there are probably exceptions is to end with the total suffix with snakecase.

So I think this would be better as `partial_success_total`